pyorbital/orbital.py

```diff
@@ -744,196 +771,230 @@ def __init__(self, orbit_elements):  # noqa: C901
                        (7.0 - 114.0 * theta2 + 395.0 * theta4) +
                        temp3 * (3.0 - 36.0 * theta2 + 49.0 * theta4))
 
-        xhdot1 = -temp1 * self.cosIO
-        self.xnodot = (xhdot1 + (0.5 * temp2 * (4.0 - 19.0 * theta2) +
+        self._xhdot1 = -temp1 * self.cosIO
+        self.xnodot = (self._xhdot1 + (0.5 * temp2 * (4.0 - 19.0 * theta2) +
                                  2.0 * temp3 * (3.0 - 7.0 * theta2)) * self.cosIO)
 
+    def _calculate_xmcof(self, coef, eeta):
         if self.eo > ECC_ALL:
-            self.xmcof = (-(2. / 3) * AE) * coef * self.bstar / eeta
-        else:
-            self.xmcof = 0.0
-
-        self.xnodcf = 3.5 * betao2 * xhdot1 * self.c1
-        self.t2cof = 1.5 * self.c1
+            return (-(2. / 3) * AE) * coef * self.bstar / eeta
+        return 0.0
 
+    def _calculate_xlcof(self):
         # Check for possible divide-by-zero for X/(1+cos(xincl)) when
         # calculating xlcof */
         temp0 = 1.0 + self.cosIO
         if np.abs(temp0) < EPS_COS:
             temp0 = np.sign(temp0) * EPS_COS
+        return 0.125 * A3OVK2 * self.sinIO * (3.0 + 5.0 * self.cosIO) / temp0
+
+    def _calculate_near_norm_parameters(self, tsi, s4):
+        c1sq = self.c1**2
+        self.d2 = 4.0 * self.aodp * tsi * c1sq
+        temp0 = self.d2 * tsi * self.c1 / 3.0
+        self.d3 = (17.0 * self.aodp + s4) * temp0
+        self.d4 = 0.5 * temp0 * self.aodp * tsi * \
+            (221.0 * self.aodp + 31.0 * s4) * self.c1
+        self.t3cof = self.d2 + 2.0 * c1sq
+        self.t4cof = 0.25 * \
+            (3.0 * self.d3 + self.c1 * (12.0 * self.d2 + 10.0 * c1sq))
+        self.t5cof = (0.2 * (3.0 * self.d4 + 12.0 * self.c1 * self.d3 + 6.0 * self.d2**2 +
+                                15.0 * c1sq * (2.0 * self.d2 + c1sq)))
 
-        self.xlcof = 0.125 * A3OVK2 * self.sinIO * \
-            (3.0 + 5.0 * self.cosIO) / temp0
+    def propagate(self, utc_time):
+        if self.mode == SGDP4_ZERO_ECC:
+            raise NotImplementedError("Mode SGDP4_ZERO_ECC not implemented")
+        elif self.mode != SGDP4_NEAR_NORM:
+            raise NotImplementedError("Deep space calculations not supported")
 
-        self.aycof = 0.25 * A3OVK2 * self.sinIO
+        return self._calculate_keplerians(utc_time)
 
-        self.cosXMO = np.cos(self.xmo)
-        self.sinXMO = np.sin(self.xmo)
-        self.delmo = (1.0 + self.eta * self.cosXMO)**3
+    def _calculate_keplerians(self, utc_time):
+        params = {"utc_time": utc_time}
+        params["ts"] = self._get_timedelta_in_minutes(params)
```

Comment:
why the switch to dict? for the rest we’re using objects, right? So how about a class for keplerian computations?

Reply:
I didn't want to put all the shared intermediate variables to attributes, which are already pretty numerous. There are also so many interconnections that I thought this was at least the easiest way to sort them out somehow.

I'll have another look tomorrow from the "add a new class" perspective.

Reply:
Ok, after "a bit" of shuffling, I did https://github.com/pytroll/pyorbital/pull/171/commits/6fb6efd6166b2a8f9be73b84b91f8a73ece222e4